File: алгоритмы/Brute_forse(Метод_Грубой_силы)/Sort_Сортировка/bubble_sort.js
// ! Сортировка пузырьком

// let arr = [4, 2, 8, 2, 2, 2, 3, 1, 55, 6, 6, 12, 3];
// let i = 0;
// let a = null;
// let b = null;
// let c = 0;

// while (c < arr.length - 1) {
//     c = 0;

//     if (arr[i] > arr[i + 1]) {
//         a = arr[i];
//         b = arr[i + 1];
//         arr[i + 1] = a;
//         arr[i] = b;
//     }

//     i = i + 1;

//     if (i === arr.length - 1) {
//         i = 0;
//     }

//     for (let ii = 0; ii < arr.length - 1; ii++) {
//         if (arr[ii] < arr[ii + 1] || arr[ii] === arr[ii + 1]) {
//             c = c + 1;
//         }
//     }
// }

// console.log(arr);

// ===========================================================================
function f(arr) {
  let bull;

  do {
    bull = false;

    for (let i = 0; i < arr.length; i++) {
      if (arr[i] > arr[i + 1]) {
        bull = true;
        [arr[i], arr[i + 1]] = [arr[i + 1], arr[i]];
      }
    }
  } while (bull);

  return arr;
}

let arr = [23, 12, 33, -11, 0, 0, -11, -3, 44, 1];
console.log(f(arr));

let arr2 = [4, 2, 8, 2, 2, 2, 3, 1, 55, 6, 6, 12, 3];

function bubbleSort(a) {
  let swapped;

  do {
    swapped = false;

    for (let i = 0, c = a.length - 1; i < c; i++) {
      if (a[i] > a[i + 1]) {
        // * более краткий способ произвести замену значений из ES6
        [a[i], a[i + 1]] = [a[i + 1], a[i]];

        swapped = true;
      }
    }
  } while (swapped);

  return a;
}

console.log(bubbleSort(arr2));

File: алгоритмы/сортировки/Brute_forse(Метод_Грубой_силы)/Sort_Сортировка/bubble_sort.go
package main

import (
	"fmt"
)

func bableSort(arr []int) []int {
	for detect := true; detect == true; {
		detect = false

		for i := 1; i < len(arr); i++ {
			if arr[i-1] > arr[i] {
				detect = true
				arr[i-1], arr[i] = arr[i], arr[i-1]
			}
		}
	}

	return arr
}

func bubbleSort(arr []int) {
	for i := 0; i < len(arr); i++ {
		for j := 0; j < len(arr)-i-1; j++ {
			if arr[j] > arr[j+1] {
				arr[j], arr[j+1] = arr[j+1], arr[j]
			}
		}
	}
}

func main() {
	fmt.Println(bableSort([]int{4, 8, 15, 1, 0, 11, 3, 6, 22, 1, 4, 8, 15}))
	fmt.Println(bableSort([]int{11, 4, 8, 15, 1, 0, 11, 3, 6, 22, 11, 1, 4, 8, 15}))
	fmt.Println(bableSort([]int{5, 4, 3, 2, 1, 0, 0, 1, 2, 3, 4, 5, 5, 4, 3, 2, 1, 0}))
	fmt.Println(bableSort([]int{5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5}))
	fmt.Println(bableSort([]int{0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5}))
	fmt.Println(bableSort([]int{0, 1, 2, 3, -6, 14, 4, 5, 0, 1, -33, 2, 3, 4, 5}))
}
